### PR TITLE
feat: show announcements to welcome page

### PR DESCRIPTION
**Pull request by daiyam**
This PR shows a list of announcements on welcome page with items found in https://github.com/VSCodium/vscodium/blob/insider/annoucements.json (depend on the quality)

@GitMensch @paulcarroty What do you think?

### CODE REVIEW

File: patches/feat-announcements.patch [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
++			return li;
++		};
++
++		if (this.announcementList) { this.announcementList.dispose(); }
++
++		const announcementList = this.announcementList = new GettingStartedIndexList({
++			title: localize('announcements', "VSCodium Announcements"),
++			klass: 'announcements',
++			limit: 5,
++			empty: $('.empty-recent', {}, localize('noAnnouncements', "There is no announcements.")),
```
Comment (GitMensch) [inline patches/feat-announcements.patch:50]: Should be "There is no current announcement."

File: patches/feat-announcements.patch [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
++			const anchor: HTMLLinkElement = $('a');
++			anchor.href = url;
++			anchor.innerText = title;
++			anchor.target = '_blank';
++			li.appendChild(anchor);
++
++			return li;
++		};
++
++		if (this.announcementList) { this.announcementList.dispose(); }
```
Comment (GitMensch) [inline patches/feat-announcements.patch:44]: If we dispose the `announcementList, then `announcementData` should be disposed, too (or otherwise both kept)

**Comment by GitMensch** [comment]
I like the general idea, but I'm not sure about how the data gets in; Wouldn't it be reasonable to at least have "static" data (that refers to the installed release), we could package a file during install and read that locally and _maybe additional_ have dynamic data?
And even with dynamic data some users will not want to have automatic file downloads (and some work without internet access in any case); so maybe cache that and get it only every N days at start [N may be 1 successful, too]?

In any case it would be good to have a screenshot how that looks like using the current announcement.json (and maybe squashing the commits before the merge).

File: patches/feat-announcements.patch (hunk 1 above)
Comment (daiyam) [reply patches/feat-announcements.patch:50]: In my test, the `buildAnnouncementList()` function was called 3 times to display the page... that why I'm caching the data. So it's fine like that.

File: patches/feat-announcements.patch (hunk 2 above)
Comment (daiyam) [reply patches/feat-announcements.patch:44]: In my test, the buildAnnouncementList() function was called 3 times to display the page... that why I'm caching the data. So it's fine like that.

**Comment by daiyam** [comment]
> Wouldn't it be reasonable to at least have "static" data (that refers to the installed release), we could package a file during install and read that locally and maybe additional have dynamic data?

Ya, but a some dynamic data is a must so we can inform the users of possible issues. So maybe a mixed approach.

> some users will not want to have automatic file downloads

I thought of that, so maybe with a new setting??

> and some work without internet access in any case

The list will be empty

> so maybe cache that and get it only every N days at start [N may be 1 successful, too]?

I think it's unnecessary.

> In any case it would be good to have a screenshot how that looks like using the current announcement.json
Screen Shot 2022-10-01 at 1 25 14 AM

> and maybe squashing the commits before the merge

Yes, I always do that when I can.

File: patches/feat-announcements.patch (hunk 1 above)
Comment (paulcarroty) [reply patches/feat-announcements.patch:50]: "There are ..."

**Comment by paulcarroty** [comment]
An option to disable dynamic data would be great.

File: patches/feat-announcements.patch (hunk 1 above)
Comment (daiyam) [reply patches/feat-announcements.patch:50]: Yep, I will fix my grammar mistake.

**Comment by daiyam** [comment]
I've added the`workbench.welcomePage.extraAnnouncements` setting to enable/disable the "dynamic" data.
The `announcements-builtin.json` file contains the announcements that are included by default.
The `announcements-extra.json` file will be fetched when the page is shown.

**Review by GitMensch — APPROVED** [review]
The new one looks very good to me, Thank you very much!

**Pull request merged**